pkg/controller/container-runtime-config/container_runtime_config_controller.go

```diff
@@ -1026,6 +1044,167 @@ func registriesConfigIgnition(templateDir string, controllerConfig *mcfgv1.Contr
+		}
+		klog.Infof("Applied default runtime MC %v on MachineConfigPool %v", managedKey, pool.Name)
+	}
+
+	// Create the config map for crio-default-container-runtime so we know that the crio-default-container-runtime MC has been created
+	if defaultContainerRuntimeCM == nil {
+		defaultContainerRuntimeCM = &v1.ConfigMap{}
+	}
+
+	defaultContainerRuntimeCM.Name = configMapName
+	defaultContainerRuntimeCM.Namespace = ctrl.namespace
+	if _, err := ctrl.kubeClient.CoreV1().ConfigMaps(ctrl.namespace).Create(context.TODO(), defaultContainerRuntimeCM, metav1.CreateOptions{}); err != nil {
+		return fmt.Errorf("error creating %s config map: %w", configMapName, err)
+	}
+	return nil
+}
+
+func (ctrl *Controller) determineContainerRuntimeForFinalRuntimeConfig(pool *mcfgv1.MachineConfigPool) (mcfgv1.ContainerRuntimeDefaultRuntime, error) {
+	// Retrieve all ContainerRuntimeConfig objects for the pool
+	ctrcfgListAll, err := ctrl.client.MachineconfigurationV1().ContainerRuntimeConfigs().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return "", fmt.Errorf("error listing container runtime configs: %w", err)
+	}
+
+	var ctrcfgList []mcfgv1.ContainerRuntimeConfig
+	for _, ctrcfg := range ctrcfgListAll.Items {
+		selector, err := metav1.LabelSelectorAsSelector(ctrcfg.Spec.MachineConfigPoolSelector)
+		if err != nil {
+			return "", fmt.Errorf("invalid label selector: %w", err)
+		}
+		if selector.Empty() || !selector.Matches(labels.Set(pool.Labels)) {
+			continue
+		}
+		ctrcfgList = append(ctrcfgList, ctrcfg)
+	}
+	// Determine the final runtime configuration
+	var currentDefaultRuntime mcfgv1.ContainerRuntimeDefaultRuntime
+	var machineConfigs []*mcfgv1.MachineConfig
+
+	for _, ctrcfg := range ctrcfgList {
+		finalizers := ctrcfg.GetFinalizers()
+		for _, finalizer := range finalizers {
+			mcName := finalizer
+			mc, err := ctrl.client.MachineconfigurationV1().MachineConfigs().Get(context.TODO(), mcName, metav1.GetOptions{})
+			if err != nil {
+				return "", fmt.Errorf("error getting MachineConfig %s: %w", mcName, err)
+			}
+			machineConfigs = append(machineConfigs, mc)
+		}
+	}
+
+	// Sort MachineConfigs in alphanumeric order
+	sort.Slice(machineConfigs, func(i, j int) bool {
+		return machineConfigs[i].Name < machineConfigs[j].Name
+	})
+
+	// Determine the final runtime configuration based on the last-applied MachineConfig
+	for i := len(machineConfigs) - 1; i >= 0; i-- {
+		mc := machineConfigs[i]
+		if mc.Spec.Config.Raw != nil {
+			ignConfig, err := ctrlcommon.ParseAndConvertConfig(mc.Spec.Config.Raw)
+			if err != nil {
+				return "", fmt.Errorf("error unmarshalling Ignition config: %w", err)
+			}
+			for _, file := range ignConfig.Storage.Files {
+				if file.Path == CRIODropInFilePathDefaultRuntime && file.Contents.Source != nil {
+					runtime := strings.TrimSpace(*file.Contents.Source)
+					if runtime == "runc" || runtime == "crun" {
```

Comment:
@MarSik this should be fine for NTO right, as it doesn't set the default runtime but just adds a new one?

Reply:
it almost feels safer not to do this check and just make sure the value isnt' empty but I am only a bit anxious about it

Reply:
xref: https://github.com/openshift/machine-config-operator/pull/4715#discussion_r1878802605

Reply:
OK let's try it out. NTO shouldn't be changing default AFAIU

Reply:
FYI these could also be changed to https://github.com/openshift/api/blob/master/machineconfiguration/v1/types.go

Reply:
> @MarSik this should be fine for NTO right, as it doesn't set the default runtime but just adds a new one?

finishing the loop: the container runtime config controller only allows values known to the api, which is runc/crun or empty, so this branch should never even be hit, and NTO shouldn't be able to use anything else